### DISCUSSION TITLE
Which iterator type instead of auto should be used?

### DISCUSSION THREAD

**Discussion by sreerajchundayil** [nlohmann/json#3039] [Q&A] [answered]
I need to understand what iterator type should I be using instead of `auto` in `nlohmann::json`.

My use case is as follows

I have merged two similar functions into one to avoid code duplication, and the only change would be at what level the iteration would happen.

```cpp
    auto itr = js["main"]["sub_main"].begin().value().begin(); 
    auto itrEnd = js["main"]["sub_main"].begin().value().end();
    if(status == false)
    {   
      itr = js["main"]["sub_main"].begin(); 
      itrEnd = js["main"]["sub_main"].end();
    }

    for(;itr!=itrEnd;itr++)
    {
     ...
    }
```
One call to function(when status == true) will have an extra key which will be removed once its call is complete and when second time the function is called(status == false) that additional key wouldn't be there.
That's the reason the redirection has been reduced inside the if condition.

The `value` will never be an array.

This code works as expected when status is true or false.
Is it safe to use `auto` like this?

How can I make two assignments into one if `status == false`?
I had to use `auto` like this because the same iterator types are used below for processing and those remaining codes are exactly same.
what should be the generic `iterator type` I should be using?

**Comment by nlohmann** [comment, marked as answer]
Of course you can use `auto` there. The type of the iterators is `json::iterator` or `json::const_iterator`, depending on the constness of `js`.
